ReactWindows/ReactNative/Bridge/DispatcherHelpers.cs

```diff
@@ -87,45 +89,157 @@ public static bool IsOnDispatcher()
+            {
+                return IsOnDispatcher();
+            }
+
+            return CoreApplication.GetCurrentView()?.Dispatcher == dispatcher;
+        }
+
+        /// <summary>
+        /// Invokes an action on the main dispatcher.
+        /// </summary>
+        /// <param name="action">The action to invoke.</param>
+        /// <param name="allowInlining">True if inlining is allowed when calling thread is on the main dispatcher.</param>
+        public static void RunOnDispatcher(DispatchedHandler action, bool allowInlining = false)
+        {
+            RunOnDispatcher(CoreDispatcherPriority.Normal, action, allowInlining);
+        }
+
+        /// <summary>
+        /// Invokes an action on the main dispatcher with custom priority.
+        /// </summary>
+        /// <param name="priority">The priority.</param>
+        /// <param name="action">The action to invoke.</param>
+        /// <param name="allowInlining">True if inlining is allowed when calling thread is on the main dispatcher.</param>
+        public static void RunOnDispatcher(CoreDispatcherPriority priority, DispatchedHandler action, bool allowInlining = false)
+        {
+            RunOnDispatcher(MainDispatcher, priority, action, allowInlining);
+        }
+
+        /// <summary>
+        /// Invokes an action on a specified dispatcher.
+        /// </summary>
+        /// <param name="dispatcher">The dispatcher.</param>
         /// <param name="action">The action to invoke.</param>
-        public static void RunOnDispatcher(DispatchedHandler action)
+        /// <param name="allowInlining">True if inlining is allowed when calling thread is on the same dispatcher as the one in the parameter.</param>
+        public static void RunOnDispatcher(CoreDispatcher dispatcher, DispatchedHandler action, bool allowInlining = false)
         {
-            RunOnDispatcher(CoreDispatcherPriority.Normal, action);
+            RunOnDispatcher(dispatcher, CoreDispatcherPriority.Normal, action, allowInlining);
         }
 
         /// <summary>
-        /// Invokes an action on the dispatcher.
+        /// Invokes an action on a specified dispatcher and priority
         /// </summary>
+        /// <param name="dispatcher">The dispatcher.</param>
         /// <param name="priority">The priority.</param>
         /// <param name="action">The action to invoke.</param>
-        public static async void RunOnDispatcher(CoreDispatcherPriority priority, DispatchedHandler action)
+        /// <param name="allowInlining">True if inlining is allowed when calling thread is on the same dispatcher as the one in the parameter.</param>
+        public static void RunOnDispatcher(CoreDispatcher dispatcher, CoreDispatcherPriority priority, DispatchedHandler action, bool allowInlining = false)
         {
-            await MainDispatcher.RunAsync(priority, action).AsTask().ConfigureAwait(false);
+            if (allowInlining && IsOnDispatcher(dispatcher))
```

Comment:
We had a problem with this pattern in UWP Classic. The problem was that when the method was used by someone who expected the calls to be serialized on the main thread, depending on where the calls will come, the order was getting messed up and as a result race conditions will occur, i.e. if we had a mixture of UI/non-UI calls these were not serialized properly due to actions not being posted always with RunAsync.

Reply:
Discussed offline with Radu.

---
In reply to: [166075753](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r166075753) [](ancestors = 166075753)